File: library.ui/src/services/AuthorService.ts
import { MessagingHelper } from "../helpers/MessagingHelper";
import axios, { Axios } from "axios";
import { ListAuthorDTO } from "../models/authors/ListAuthorDTO";
import { CreateAuthorDTO } from "../models/authors/CreateAuthorDTO";
import { PaginatedList } from "../helpers/PaginatedList";
import { SearchDTO } from "../helpers/SeacrhDTO";
import { Parameter } from "../helpers/Parameter";
import { AuthorDTO } from "../models/authors/AuthorDTO";
import { EditBookDTO } from "../models/books/EditBookDTO";
import { EditAuthorDTO } from "../models/authors/EditAuthor";
import { APIService } from "./APIService";

export class AuthorService{
    async GetAll(currentPage: number, PageSize: number, searchParameter: Parameter[], OrderByParameter: Parameter[] ) : Promise<PaginatedList<ListAuthorDTO>>{
        try{
            const search = new SearchDTO(currentPage+1, PageSize, searchParameter, OrderByParameter)
            var response = await APIService.Axios().post(`${APIService.GetURL()}/Author/getAll`,{
                ...search
            },{
                headers:{
                    Accept: "application/json",
                    "Content-Type": "application/json",
                    "Authorization": "Bearer " + APIService.GetToken()
                }
                
            });
            
            return response.data
            
  
        }catch(error){
           return new PaginatedList<ListAuthorDTO>(false, "", "", [], 0, false, true);
        }
    }

    async GetAuthors() : Promise<MessagingHelper<ListAuthorDTO[]>>{
        try{
            
            var response = await APIService.Axios().get(`${APIService.GetURL()}/Author/getAuthors`,{
                headers:{
                    Accept: "application/json",
                    "Content-Type": "application/json",
                    "Authorization": "Bearer " + APIService.GetToken()
                }
            });
            return response.data
        }catch(error){
            return new MessagingHelper<ListAuthorDTO[]>(false, "erro ao buscar os autores", []);
        }
    }

    async Delete(id: number) : Promise<MessagingHelper<null>>{
        try{
            var response = await APIService.Axios().delete(`${APIService.GetURL()}/Author/delete/`+id,{
                headers:{
                    Accept: "application/json",
                    "Content-Type": "application/json",
                    "Authorization": "Bearer " + APIService.GetToken()
                }
            });
            return response.data;
        }catch (error){
            return new MessagingHelper<null>(false,"Erro ao eliminar o author", null);
        }

    } 

    async Create(author: CreateAuthorDTO) : Promise<MessagingHelper<null>>{
        try{
            var response = await APIService.Axios().post(`${APIService.GetURL()}/Author/create`, {
                ...author
            },{
                    headers:{
                        Accept: "application/json",
                        "Content-Type": "application/json",
                        "Authorization": "Bearer " + APIService.GetToken()
                    }
            });
            return response.data;
        }catch(error){
            return new MessagingHelper<null>(false, "Erro ao criar o author", null);
        }
    }

    async GetById(id: number) : Promise<MessagingHelper<AuthorDTO | null>>{
        try{
            var response = await APIService.Axios().get(`${APIService.GetURL()}/Author/` + id,{
                headers:{
                    Accept: "application/json",
                    "Content-Type": "application/json",
                    "Authorization": "Bearer " + APIService.GetToken()
                }
            });
            return response.data;
        }catch(error){
            return new MessagingHelper<AuthorDTO | null>(false, "Erro ao pesquisar o autor", null);
        }
    }

    async Update(author: EditAuthorDTO) : Promise<MessagingHelper<AuthorDTO | null>>{
        try{
            var response = await APIService.Axios().post(`${APIService.GetURL()}/Author/update`, {
                ...author
            },{
                headers:{
                    Accept: "application/json",
                    "Content-Type": "application/json",
                    "Authorization": "Bearer " + APIService.GetToken()
                }
            });
            return response.data;
        }catch(error){
            return new MessagingHelper<AuthorDTO | null>(false, "Erro ao editar o autor", null);
        }
    }
}

File: Library.DAL/Services/AuthorService.cs
using Supermarket.Data.Interfaces.Repository;
using Supermarket.Data.Interfaces.Services;
using Supermarket.Data.Models.Authors;
using Supermarket.Data.Models.Helper;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Library.DAL.Services
{
    public class AuthorService : IAuthorService
    {
        private readonly IBookRepository _bookRepository;
        private readonly IAuthorRepository _authorRepository;
        public AuthorService(IBookRepository bookRepository, IAuthorRepository authorRepository)
        {
            _bookRepository = bookRepository;
            _authorRepository = authorRepository;
        }
        public async Task<MessageHelper> Create(CreateAuthorDTO createAuthor)
        {
            MessageHelper result = new MessageHelper();
            try
            {
                CreateAuthorDTOValidator validator = new CreateAuthorDTOValidator();
                var responseValidate = await validator.ValidateAsync(createAuthor);
                if (responseValidate == null || responseValidate.IsValid == false) {
                    if (responseValidate == null) {
                        result.Message = "Erro ao validar a informacao";
                        return result;
                    }
                    result.Message = responseValidate.Errors.FirstOrDefault()!.ErrorMessage;
                    return result;
                }
                var AlreadyHaveAuthor = _authorRepository.GetByName(createAuthor.Name);
                if (AlreadyHaveAuthor != null) {
                    result.Message = "Ja exiiste este autor";
                    return result;
                }

                var author = await _authorRepository.Create(createAuthor.ToEntity());

                if (author == null) {
                    result.Sucess = false;
                    result.Message = "Nao foi possivel criar o autor";
                    return result;
                }
                result.Sucess = true;
                result.Message = "Autor criado com sucesso";
            }
            catch (Exception ex) {
                result.Sucess = false;
                result.Message = ex.Message;
                return result;
            }
            return result;

        }

        public async Task<MessageHelper> Delete(int id)
        {
            MessageHelper result = new MessageHelper();
            try {
                var haveAuthor = _authorRepository.GetById(id).Result;

                if (haveAuthor == null) {
                    result.Sucess = false;
                    result.Message = "Este Autor nao existe";
                    return result;
                }

                var deletebook = await _authorRepository.Delete(haveAuthor);

                if (deletebook == false) {
                    result.Sucess = false;
                    result.Message = "Nao foi possivel eliminar o Autor";
                    return result;
                }

                result.Sucess= true;
                result.Message = "Autor eliminado com sucesso";
            } catch (Exception ex) {
                result.Sucess = false;
                result.Message = ex.Message;
            }
            return result;
        }

        public async Task<PaginateList<ListAuthorDTO>> GetAll(SearchDTO search)
        {
            PaginateList<ListAuthorDTO> result = new PaginateList<ListAuthorDTO>();
            try
            {
                if (search.PageSize > 100)
                {
                    search.PageSize = 100;
                }

                if (search.PageSize <= 0)
                {
                    search.PageSize = 1;
                }

                if (search.CurrentPage <= 0)
                {
                    search.CurrentPage = 1;
                }

                var responseRepository = await _authorRepository.GetAll(search.SearchBy, search.OrderBy, search.CurrentPage, search.PageSize);
                result.Items = responseRepository.Items.Select(t => new ListAuthorDTO(t)).ToList();
                result.Success = true;
                result.TotalRecords = responseRepository.TotalRecords;
                result.CurrentPage = responseRepository.CurrentPage;
                result.PageSize = responseRepository.PageSize;


                return result;

            }
            catch (Exception ex)
            {
                result.Success = false;
                return result;
            }
        }

        public async Task<MessageHelper<List<ListAuthorDTO>>> GetAuthors()
        {
            MessageHelper<List<ListAuthorDTO>> result = new MessageHelper<List<ListAuthorDTO>>();
            try
            {   
                var authors = await _authorRepository.GetAuthors();
                if (authors == null) {
                    result.Message = "Erro";
                    result.Sucess = false;
                    return result;
                }
                result.obj = authors.Select(t => new ListAuthorDTO(t)).ToList();
                result.Sucess = true;

            }
            catch (Exception ex) {
                   result.Message = ex.Message;
                   result.Sucess = false;
                   
                   return result;
            }
            return result;
        }

        public async Task<MessageHelper<AuthorDTO>> GetById(int authorid)
        {
            MessageHelper<AuthorDTO> result = new MessageHelper<AuthorDTO>();
            try
            {
                var author = await _authorRepository.GetById(authorid);

                if (author == null)
                {
                    result.Sucess = false;
                    result.Message = "Nao foi possivel encontrar este autor";
                    return result;
                }
                result.obj = new AuthorDTO(author);
                result.Sucess = true;
            }
            catch (Exception ex) {
                result.Sucess= false;
                result.Message = ex.Message;
                return result;
            }
            return result;

            
        }

        public async Task<MessageHelper<AuthorDTO>> Update(EditAuthorDTO editAuthor)
        {
            MessageHelper<AuthorDTO> result = new();
            try
            {
                EditAuthorDTOValidator validator = new EditAuthorDTOValidator();
                var responseValidate = validator.Validate(editAuthor);
                if (responseValidate.IsValid == false || responseValidate == null) {

                    if (responseValidate == null)
                    {
                        result.Sucess = false;
                        result.Message = "Erro ao validar a informacao";
                        return result;
                    }

                    result.Message = responseValidate.Errors.FirstOrDefault()!.ErrorMessage;
                    return result;
                }

                var author = await _authorRepository.GetById(editAuthor.AuthorId);
                if (author == null) {
                    result.Sucess = false;
                    result.Message = "Nao existe este Autor";
                    return result;
                }
                author.Name = editAuthor.Name;
                author.BirthDate = editAuthor.BirthDate;
                author.DeadDate = editAuthor.DeadDate;

                author = await _authorRepository.Update(author);

                result.Sucess = true;
                result.obj = new AuthorDTO(author);
                
            }
            catch (Exception ex) {
                result.Sucess = false;
                result.Message = ex.Message;
            }
            return result;
        }
    }
}
